planet/api/http.py

```diff
@@ -13,135 +13,176 @@
-    strip Authorization headers from any redirect requests that resolve to a
-    new domain. Instead, we'll keep headers if the redirect is a subdomain
-    and if not, extract the api-key from the header and add it to the url
-    as a parameter.
-    '''
-    def rebuild_auth(self, prepared_request, response):
-        existing_auth = prepared_request.headers.get('Authorization', None)
-        if existing_auth:
-            orig = response.request.url
-            redir = prepared_request.url
-            if not self._is_subdomain_of_tld(orig, redir):
-                prepared_request.headers.pop('Authorization')
-                key = re.match(r'api-key (\S+)', existing_auth)
-                if key:
-                    prepared_request.prepare_url(
-                        prepared_request.url, {
-                            'api_key': key.group(1)
-                        }
-                    )
+        http_resp = await self._client.send(request.http_request,
+                                            stream=stream)
+        return models.Response(request, http_resp)
+
+    def stream(self, request):
+        '''Submit a request and get the response as a stream context manager.
+
+        :param request: Request to submit
+        :type request: planet.api.models.Request
+        :returns: Context manager providing the body as a stream.
+        :rtype: APlanetStream
+        '''
+        return APlanetStream(
+            session=self,
+            request=request
+        )
 
     @staticmethod
-    def _is_subdomain_of_tld(url1, url2):
-        orig_host = urlparse(url1).hostname
-        re_host = urlparse(url2).hostname
-        return orig_host.split('.')[-2:] == re_host.split('.')[-2:]
+    def _get_user_agent():
+        return 'planet-client-python/' + __version__
+
+    @staticmethod
+    async def _log_request(request):
+        LOGGER.info(f'{request.method} {request.url} - Sent')
+
+    @staticmethod
+    async def _log_response(response):
+        request = response.request
+        LOGGER.info(
+            f'{request.method} {request.url} - '
+            f'Status {response.status_code}')
+
+    @staticmethod
+    async def _raise_for_status(response):
+        # TODO: consider using http_response.reason_phrase
+        status = response.status_code
+
+        if status < 300:
+            return
+
+        exception = {
+            400: exceptions.BadQuery,
+            401: exceptions.InvalidAPIKey,
+            403: exceptions.NoPermission,
+            404: exceptions.MissingResource,
+            429: exceptions.TooManyRequests,
+            500: exceptions.ServerError
+        }.get(status, None)
+
+        try:
+            msg = response.text
+        except httpx.ResponseNotRead:
+            await response.aread()
+            msg = response.text
+
+        # differentiate between over quota and rate-limiting
+        if status == 429 and 'quota' in msg.lower():
```

Comment:
I believe this branch can be eliminated with something like `lambda m: exceptions.OverQuota(m) if 'quota' in m.lower() else exceptions.TooManyRequests(m)` in the key lookup for `exception` in the anonymous `dict`.

Reply:
We can revisit this in future development